### ISSUE TITLE
Create or update of a topic that already exists fails if the connection is SSL

### ISSUE THREAD

**Issue by justCatchingRye** [StephenSorriaux/ansible-kafka-admin#53]
## Expected Behavior

Create of a topic that already exists returns ok
Update of a topic that already exists returns changed

## Actual Behavior

Play fails and exits with msg:

> Error while initializing Zookeeper client : Connection time-out. Is your Zookeeper server available and running on...

Note the playbook runs the first time, but fails the second time. So, this means that you can create topics, but you can't add topics to the playbook and you can't update topics.

## Play to Reproduce the Problem

```
---
- hosts: bastion_branch
  vars:
      zookeeper: 'REMOVED:2181'
      bootstrap_servers: 'REMOVED:9094'
    api_version: '2.2.1'
    security_protocol: 'SSL'
    ssl_supported_protocols: 'TLSv1.2'
    ssl_check_hostname: False
    ssl_certfile: "{{cert_pem}}"
    ssl_keyfile: "{{key_pem}}"
    topics:
      - "{{env}}.{{product}}.topic"
  tasks:
  - name: Create Topic
    with_items: "{{topics}}"
    kafka_lib:
      api_version: "{{api_version}}"
      security_protocol: "{{security_protocol}}"
      ssl_supported_protocols: "{{ssl_supported_protocols}}"
      zookeeper: "{{zookeeper}}"
      zookeeper_ssl_keyfile: "{{ssl_keyfile}}"
      zookeeper_ssl_certfile: "{{ssl_certfile}}"
      zookeeper_ssl_check_hostname: "{{ssl_check_hostname}}"
      bootstrap_servers: "{{bootstrap_servers}}"
      ssl_check_hostname: "{{ssl_check_hostname}}"
      ssl_certfile: "{{ssl_certfile}}"
      ssl_keyfile: "{{ssl_keyfile}}"
      resource: 'topic'
      name: "{{item}}"
      partitions: 1
      replica_factor: 2
      options:
        retention.ms: -1
      state: 'present'
```

## Logs from the play with Ansible in debug mode

On the first run, it works:

```
changed: [REMOVED.com] => (item=env.product.test) => {
    "ansible_loop_var": "item",
    "changed": true,
    "invocation": {
        "module_args": {
            "acl_host": "*",
            "acl_operation": null,
            "acl_pattern_type": "literal",
            "acl_permission": "allow",
            "acl_principal": null,
            "acl_resource_type": "topic",
            "api_version": "2.2.1",
            "bootstrap_servers": "REMOVED:9094,REMOVED:9094,REMOVED:9094",
            "name": "env.product.test",
            "options": {
                "retention.ms": -1
            },
            "partitions": 1,
            "replica_factor": 2,
            "resource": "topic",
            "sasl_kerberos_service_name": null,
            "sasl_mechanism": "PLAIN",
            "sasl_plain_password": null,
            "sasl_plain_username": null,
            "security_protocol": "SSL",
            "ssl_cafile": null,
            "ssl_certfile": "REMOVED.crt",
            "ssl_check_hostname": false,
            "ssl_ciphers": null,
            "ssl_crlfile": null,
            "ssl_keyfile": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "ssl_password": null,
            "ssl_supported_protocols": [
                "TLSv1.2"
            ],
            "state": "present",
            "zookeeper": "REMOVED:2181,REMOVED:2181,REMOVED:2181",
            "zookeeper_auth_scheme": "digest",
            "zookeeper_auth_value": "",
            "zookeeper_max_retries": 5,
            "zookeeper_sleep_time": 5,
            "zookeeper_ssl_cafile": null,
            "zookeeper_ssl_certfile": "REMOVED.crt",
            "zookeeper_ssl_check_hostname": false,
            "zookeeper_ssl_keyfile": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "zookeeper_ssl_password": null
        }
    },
    "item": "env.product.test",
    "msg": "topic 'env.product.test': successfully created."
```

When re-run it fails, it states there is a connection timeout, but that doesn't seem likely:

```
The full traceback is:
WARNING: The below traceback may *not* be related to the actual failure.
  File "/tmp/ansible_kafka_lib_payload_CzGeY8/__main__.py", line 1758, in main
    verify_certs=zookeeper_ssl_check_hostname
  File "/tmp/ansible_kafka_lib_payload_CzGeY8/__main__.py", line 879, in init_zk_client
    self.zk_client.start()
  File "/home/REMOVED /.local/lib/python2.7/site-packages/kazoo/client.py", line 582, in start
    raise self.handler.timeout_exception("Connection time-out")
failed: [REMOVED.com] (item=env.product.test) => {
    "ansible_loop_var": "item",
    "changed": false,
    "invocation": {
        "module_args": {
            "acl_host": "*",
            "acl_operation": null,
            "acl_pattern_type": "literal",
            "acl_permission": "allow",
            "acl_principal": null,
            "acl_resource_type": "topic",
            "api_version": "2.2.1",
            "bootstrap_servers": "REMOVED:9094,REMOVED:9094,REMOVED:9094",
            "name": "env.product.test",
            "options": {
                "retention.ms": -1
            },
            "partitions": 1,
            "replica_factor": 2,
            "resource": "topic",
            "sasl_kerberos_service_name": null,
            "sasl_mechanism": "PLAIN",
            "sasl_plain_password": null,
            "sasl_plain_username": null,
            "security_protocol": "SSL",
            "ssl_cafile": null,
            "ssl_certfile": "REMOVED.crt",
            "ssl_check_hostname": false,
            "ssl_ciphers": null,
            "ssl_crlfile": null,
            "ssl_keyfile": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "ssl_password": null,
            "ssl_supported_protocols": [
                "TLSv1.2"
            ],
            "state": "present",
            "zookeeper": "REMOVED:2181,REMOVED:2181,REMOVED:2181",
            "zookeeper_auth_scheme": "digest",
            "zookeeper_auth_value": "",
            "zookeeper_max_retries": 5,
            "zookeeper_sleep_time": 5,
            "zookeeper_ssl_cafile": null,
            "zookeeper_ssl_certfile": "REMOVED.crt",
            "zookeeper_ssl_check_hostname": false,
            "zookeeper_ssl_keyfile": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "zookeeper_ssl_password": null
        }
    },
    "item": "env.product.test",
    "msg": "Error while initializing Zookeeper client : Connection time-out. Is your Zookeeper server available and running on 'REMOVED:2181,REMOVED:2181,REMOVED:2181'?"
}

PLAY RECAP **************************************************************************************************************************************************************************
ansible_python_interpreter=/usr/bin/python : ok=0    changed=0    unreachable=1    failed=0    skipped=0    rescued=0    ignored=0   
REMOVED.com : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

## Specifications

  - Library version: 
  - Result of `pip list` command: 
```
Package                          Version
-------------------------------- -------
ansible                          2.6.19
awscli                           1.17.6
Babel                            0.9.6
backports.ssl-match-hostname     3.5.0.1
boto                             2.49.0
boto3                            1.10.37
botocore                         1.13.50
cffi                             1.6.0
chardet                          2.2.1
cloud-init                       18.5
colorama                         0.4.1
configobj                        4.7.2
cryptography                     1.7.2
decorator                        3.4.0
docutils                         0.15.2
enum34                           1.0.4
ethtool                          0.8
futures                          3.1.1
httplib2                         0.9.2
idna                             2.4
iniparse                         0.4
ipaddr                           2.1.11
ipaddress                        1.0.16
IPy                              0.75
javapackages                     1.0.0
Jinja2                           2.7.2
jmespath                         0.9.0
jsonpatch                        1.2
jsonpointer                      1.9
kafka-python                     1.4.4
kazoo                            2.6.1
kitchen                          1.1.1
lxml                             3.2.1
M2Crypto                         0.21.1
Magic-file-extensions            0.2
MarkupSafe                       0.11
paramiko                         2.1.1
passlib                          1.6.5
pciutils                         1.7.3
perf                             0.1
pip                              20.0.1
ply                              3.4
policycoreutils-default-encoding 0.1
prettytable                      0.7.2
pure-sasl                        0.5.1
pyasn1                           0.1.9
pycparser                        2.14
pycrypto                         2.6.1
pycurl                           7.19.0
pygobject                        3.22.0
pygpgme                          0.3
pyinotify                        0.9.4
pyliblzma                        0.5.3
pyOpenSSL                        0.13.1
pyserial                         2.6
PySocks                          1.5.7
python-dateutil                  2.8.0
python-dmidecode                 3.10.13
python-linux-procfs              0.4.9
pytoml                           0.1.14
pyudev                           0.15
pyxattr                          0.5.1
PyYAML                           3.10
registries                       0.1
requests                         2.6.0
rhnlib                           2.5.65
rsa                              3.4.2
s3transfer                       0.2.1
schedutils                       0.4
seobject                         0.1
sepolicy                         1.1
setuptools                       0.9.8
shyaml                           0.6.1
six                              1.14.0
slip                             0.4.0
slip.dbus                        0.4.0
subscription-manager             1.24.13
syspurpose                       1.24.13
urlgrabber                       3.10
urllib3                          1.21.1
wheel                            0.33.6
yum-metadata-parser              1.1.4
```
  - Kafka version: 2.2.1
  - Python version: 2.7.5, 3.7.6 (tried both)
  - OS: Red Hat Enterprise Linux Server release 7.7 (Maipo)

**Comment by StephenSorriaux** [comment]
Hey,

Thanks for this issue. I think it is related to https://github.com/python-zk/kazoo/issues/587 which I need to link into.

**Comment by StephenSorriaux** [comment]
Hey @justCatchingRye,

Can you please provide more information your ZK configuration? I am trying to reproduce the issue.

Thank you.

**Comment by justCatchingRye** [comment]
Hey StephenSorriaux,

Many thanks for following-up on this. You prompted me to take a closer look at our zookeeper setup and I've found that the issue I was seeing was actually mis-configuration on my side, and so the issue can be closed I think.

We are using AWS MSK (Managed Streaming for Kafka) and I've realised that although our connections to brokers are using TLS, our connections to zookeeper aren't. This is a bit of a surprise, but I realised that when we talk to zookeeper directly we don't actually provide a certificate for the connection. Furthermore, we are connecting to port 2181 and it appears this is the plaintext port.

So, I've resolved the issue in the playbook by removing these lines:

zookeeper_ssl_check_hostname: "{{ssl_check_hostname}}"
zookeeper_ssl_certfile: "{{ssl_certfile}}"
zookeeper_ssl_keyfile: "{{ssl_keyfile}}"

In my problem description I said that creating topics worked, then updates failed. So, this makes sense if I had the zookeeper connection setup using vars that it shouldn't have had, as zookeeper is used in cases where a topic already exists (e.g. an update action). It all makes sense now :)

Thank you very much looking into this and apologies for any time wasted looking into it.
